### PR TITLE
Remove thread and guard condition from Timer

### PR DESCRIPTION
**Pull request by jacquelinekay**
Connects to #6 

Depends on ros2/rclcpp/pull/66 and related PRs

Removes the 1 thread, 1 guard condition per timer scheme. Instead, the executor checks if a timer callback is eligible to be executed in `get_next_timer` based on the timer period and the last time the callback was triggered.

Initial CI run builds for all, and tests pass for all but Windows:

http://ci.ros2.org/job/ros2_batch_ci_linux/104/
http://ci.ros2.org/job/ros2_batch_ci_osx/45/
http://ci.ros2.org/job/ros2_batch_ci_windows/78/

I'm open to feedback regarding any tests that should be added to check timing correctness.

### CODE REVIEW

**Comment by tfoote** [comment]
+1

**Comment by jacquelinekay** [comment]
more CI after rebasing:

http://ci.ros2.org/job/ros2_batch_ci_linux/120/
http://ci.ros2.org/job/ros2_batch_ci_osx/62
http://ci.ros2.org/job/ros2_batch_ci_windows/89/

OSX has one test failure, which perhaps will be fixed by: https://github.com/ros2/system_tests/pull/23

**Comment by esteve** [comment]
+1

**Comment by jacquelinekay** [comment]
I'd like @wjwwood to comment on this before I consider merging since he suggested these changes originally, so I'll wait until he's back on the grid.

File: rclcpp/include/rclcpp/executor.hpp [hunk 1]
```diff
@@ -596,31 +548,25 @@ class Executor
...
         }
-        if (!group->can_be_taken_from_.load()) {
-          // Group is mutually exclusive and is being used, so skip it for now
-          // Leave it to be checked next time, but continue searching
-          continue;
+        for (auto timer : group->timer_ptrs_) {
+          if (timer->check_and_trigger()) {
+            any_exec->timer = timer;
+            any_exec->callback_group = group;
+            any_exec->node = get_node_by_group(group);
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: Can't you just use `node` here?

**Comment by wjwwood** [comment]
Don't you need to make use of the timeout to wait before merging this?

File: rclcpp/include/rclcpp/executor.hpp (hunk 1 above)
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/executor.hpp]: I didn't write this originally, but sure, should work.

**Comment by dirk-thomas** [comment]
LGTM.

I do have a test locally for timers. I will update it after the merge and make a PR to add it to `test_rclcpp`.

**Comment by jacquelinekay** [comment]
You mean the changes made in the `wait_timeout` branches? I'm not sure how
to incorporate those changes, maybe you can clarify the expected behavior
for me. Should the timeout pased to `rmw_wait` limit the size of the
timeouts of the node timers?

**Comment by jacquelinekay** [comment]
@dirk-thomas thanks, a timers test would be great.

**Comment by wjwwood** [comment]
As far as I can tell, this code as is will never wake up from a blocking call to spin in order to process timers. Since there is no longer a guard condition which wakes up rmw_wait then calls to rmw_wait must be setup to timeout in time for the next timer. In order to do that you need the ability to do a timed call to rmw_wait.

**Comment by dirk-thomas** [comment]
Can we cover such a case with a test then - to ensure it works after the patch?

**Comment by wjwwood** [comment]
I can look at your test and try to modify it. The important part is to test it without using subscriptions (so that the wait is only interrupted by signals and timers). And make sure that the timer is fired multiple times and that a blocking spin is used. I think the "simplest" way to test this is to:

- Create a timer with a callback that increments a global variable.
- Call `spin` in a separate thread.
- Wait enough time for several timer callbacks to have fired.
- Call `rclcpp::shutdown`.
- Join the spin thread.
- Check the number of times the timer callback fired and assert that it fired a few times.

I can try to put together a test as described, but I need to get my other pull request open first, so someone else might want to take a stab it as to not slow this pr down.

**Comment by dirk-thomas** [comment]
I have created ros2/system_tests#25 which adds a simple test for timers. I does roughly what @wjwwood  described just without an extra thread and shutdown call. @jacquelinekay Can you try the test with your branch?

**Comment by jacquelinekay** [comment]
To fix the problem @wjwwood pointed out a few days ago, I merged this with the `wait_timeout` branch to get an `rmw_wait` call with a timeout, which has a few major API changes that cause the test to not compile. I'll make a branch of system_tests with the changes and test this branch with it.

To simplify things, we should probably focus on the review process for the `wait_timeout` pull requests before progressing with this one.

**Comment by jacquelinekay** [comment]
The timer gtest passes locally on my machine. Let's see how Jenkins goes. Will run Windows when it's fixed.

http://ci.ros2.org/job/ros2_batch_ci_linux/158/
http://ci.ros2.org/job/ros2_batch_ci_osx/86/

**Comment by jacquelinekay** [comment]
The same 4 tests fail in both jobs (missing result) and I cannot reproduce the failures locally, either by running them individually or "make test" for test_rclcpp.

**Comment by dirk-thomas** [comment]
The console output will show what the problem for these four tests is. They all timeout.

It could be that the timeout is just very close to the actual runtime of the test and therefore results in a flaky test. But looking at the log it seems more likely that your changes are affecting the behavior:

    Expected: (nonblocking_diff) < (tolerance), actual: 8-byte object <0C-CB 03-00 01-00 00-00> vs 8-byte object <0F-00 00-00 00-00 00-00>

**Comment by jacquelinekay** [comment]
I corrected some boolean logic that should fix the tests.

http://ci.ros2.org/job/ros2_batch_ci_linux/159/
http://ci.ros2.org/job/ros2_batch_ci_osx/87/

**Comment by tfoote** [comment]
+1

**Comment by esteve** [comment]
+1

File: rclcpp/include/rclcpp/executor.hpp [hunk 2]
```diff
@@ -598,34 +562,58 @@ class Executor
...
+      }
+    }
+  }
+
+  std::chrono::nanoseconds
+  get_earliest_timer()
+  {
+    std::chrono::nanoseconds latest = std::chrono::nanoseconds::max();
+    bool timers_empty = true;
+    for (auto weak_node : weak_nodes_) {
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: Might want to use `auto & weak_node`. The compiler might be smart enough to avoid extra work here, but we can help it out.

File: rclcpp/include/rclcpp/executor.hpp [hunk 3]
```diff
@@ -598,34 +562,58 @@ class Executor
...
+      auto node = weak_node.lock();
+      if (!node) {
+        continue;
+      }
+      for (auto weak_group : node->callback_groups_) {
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: Same here.

File: rclcpp/include/rclcpp/executor.hpp [hunk 4]
```diff
@@ -598,34 +562,58 @@ class Executor
...
+        if (!group || !group->can_be_taken_from_.load()) {
           continue;
         }
-        // Otherwise it is safe to set and return the any_exec
-        any_exec->timer = timer;
-        any_exec->callback_group = group;
-        any_exec->node = get_node_by_group(group);
-        guard_condition_handles_.erase(it++);
-        return;
+        for (auto timer_ref : group->timer_ptrs_) {
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: Same here, and anywhere else you're using for-range loop.

Comment (jacquelinekay) [inline rclcpp/include/rclcpp/executor.hpp]: there are several instances in executor where references are not being used
in range-based for loops, should I go ahead and correct them in one fell
swoop?

File: rclcpp/include/rclcpp/timer.hpp [hunk 5]
```diff
@@ -106,29 +95,49 @@ class GenericTimer : public TimerBase
...
+  // return: true to trigger callback on the next "execute_timer" call in executor
+  bool
+  check_and_trigger()
+  {
+    if (canceled_) {
+      return false;
+    }
+    if (Clock::now() < last_triggered_time_) {
+      // last trigger time was in the future, this should never happen
+      throw std::runtime_error("Received invalid last triggered time");
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/timer.hpp]: It is absolutely possible that the clock jumps backward and should be handled more gracefully.

File: rclcpp/include/rclcpp/executor.hpp [hunk 6]
```diff
@@ -368,21 +360,25 @@ class Executor
...
-    }
 
     rmw_time_t * wait_timeout = NULL;
     rmw_time_t rmw_timeout;
 
-    if (timeout >= std::chrono::duration<int64_t, T>::zero()) {
+    auto next_timer_duration = get_earliest_timer();
+    /* If the next timer timeout must preempt the requested timeout
+       or if the requested timeout blocks forever, and there exists a valid timer,
+       replace the requested timeout with the next timeout. */
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/executor.hpp]: Please use multiple single line comments.

Same below.

File: rclcpp/include/rclcpp/timer.hpp [hunk 7]
```diff
@@ -106,29 +95,49 @@ class GenericTimer : public TimerBase
...
+      loop_rate_.period())
+    {
+      last_triggered_time_ = Clock::now();
+      return true;
+    }
+    return false;
+  }
+
+  std::chrono::nanoseconds
+  time_to_trigger()
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/timer.hpp]: Minor comment, `time_to_trigger` sounds like the absolute (datetime) time on which to trigger. Maybe `time_until_trigger` or `time_till_trigger` better indicates that this is a duration until the next trigger.

File: rclcpp/include/rclcpp/timer.hpp (hunk 5 above)
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/timer.hpp]: It's true, the timer class is templated on the `Clock` time so it's not required that the clock is monotonic :/ If the clock jumps backward before the last triggered time I will simply return false out of this function and not change the state of the timer. I'm not sure if there's much else I can do.

**Comment by dirk-thomas** [comment]
I am not sure which commits are different since the last review but looks still good to me.

File: rclcpp/include/rclcpp/executor.hpp [hunk 8]
```diff
@@ -351,8 +344,7 @@ class Executor
 
     // Use the number of guard conditions to allocate memory in the handles
     // Add 2 to the number for the ctrl-c guard cond and the executor's
-    size_t start_of_timer_guard_conds = 2;
-    size_t number_of_guard_conds = timers.size() + start_of_timer_guard_conds;
+    size_t number_of_guard_conds = 2;
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: Can you update the comment to reflect this new value?

File: rclcpp/include/rclcpp/executor.hpp [hunk 9]
```diff
@@ -368,21 +360,25 @@ class Executor
...
     rmw_time_t rmw_timeout;
 
-    if (timeout >= std::chrono::duration<int64_t, T>::zero()) {
+    auto next_timer_duration = get_earliest_timer();
+    /* If the next timer timeout must preempt the requested timeout
+       or if the requested timeout blocks forever, and there exists a valid timer,
+       replace the requested timeout with the next timeout. */
+    if ( (next_timer_duration < timeout || timeout <
+      std::chrono::duration<int64_t,
+      T>::zero()) && next_timer_duration >= std::chrono::nanoseconds::zero())
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: If you want to use whitespace, e.g. `if ( (...` to separate things out, please just make them symmetric. So, put a whitespace before the final `)` since there is one proceeding the first `(`.

In general these multiline conditionals are pretty hard to read, and there not much you can do about that. Maybe using some more vertical space would make it easier, maybe like this:

```c++
    if (
      (
        next_timer_duration < timeout ||
        timeout < std::chrono::duration<int64_t, T>::zero()
      ) &&
      next_timer_duration >= std::chrono::nanoseconds::zero()
    ) {
      // ...
    }
```

Honestly my suggestion doesn't look much better, but maybe it will inspire you to come up with a different solution.

Comment (jacquelinekay) [inline rclcpp/include/rclcpp/executor.hpp]: Uncrustify is always screwing up my ridiculous line lengths. I think I may instead put one or more of these clauses into bools, which makes this conditional more readable and may eliminate the whitespace problem...

File: rclcpp/include/rclcpp/executor.hpp (hunk 4 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: Well, you need to consider what the implications are, but yes usually the auto reference is a safe optimization to do. I'd say just update the ones you're adding and anything related to your changes. If you want to audit the rest, I'd do that in a separate pr so we can more easily review each case.
